src/postgraphile/postgraphile.ts

```diff
@@ -104,11 +93,23 @@ export function getPostgraphileSchemaBuilder<
     while (true) {
       try {
         if (options.watchPg) {
-          await watchPostGraphileSchema(pgPool, pgSchemas, options, newSchema => {
-            gqlSchema = newSchema;
-            _emitter.emit('schemas:changed');
-            exportGqlSchema(gqlSchema);
+          const releaseWatchFnPromise = watchPostGraphileSchema(
+            pgPool,
+            pgSchemas,
+            options,
+            newSchema => {
+              gqlSchema = newSchema;
+              _emitter.emit('schemas:changed');
+              exportGqlSchema(gqlSchema);
+            },
+          );
+          // We must register the shutdown action immediately to avoid a race condition.
+          shutdownActions.add(async () => {
+            const releaseWatchFn = await releaseWatchFnPromise;
+            releaseWatchFn();
```

Comment:
We should return this, I think it's a promise in itself.

```suggestion
            return releaseWatchFn();
```